helm/nginx-ingress-controller-app/values.yaml

```diff
@@ -130,6 +130,39 @@ controller:
     # The name tha the RBAC role will get.
     name: nginx-ingress-role
 
+  # controller.livenessProbe
+  # Liveness probe configuration
+  # Ref: https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#container-probes
+  livenessProbe:
+    enabled: true
+    failureThreshold: 30
+    initialDelaySeconds: 10
+    periodSeconds: 10
+    successThreshold: 1
+    timeoutSeconds: 5
```

Comment:
liveness probe:
- failureThreshold default was changed from 3 to 30
- timeoutSeconds default was changed from 1 to 5
- periodSeconds (frequency) is left unchanged, stays same as for readiness probe.

With these settings we tolerate liveness probe to fail up to 5min before considering container as failed. Only then kubelet will restart the container.
Readiness probe configuration is left unchanged (if I'm not mistaken, it's on k8s defaults).

When nginx replica gets overloaded (/w max concurrent connections limit reached),
- readiness probe failure threshold will be reached first (before liveness does)
- as readiness threshold gets reached replica will be removed from load balancing
  - replica won't receive additional traffic
  - replica will have time to process existing requests it received
  - at least some existing requests are expected to be processed within 5min, and connection capacity freed up
- as capacity is freed up, liveness and readiness checks are expected to be successful
- with readiness probe passing again, more traffic will be allowed through to replica

As consequence nginx replica recovers faster from overload and can serve more capacity since
- it will stop receiving additional traffic, and
- replica won't be killed by kubelet as often as before - liveness probe will have higher chance to pass even throughout overloaded periods.

With this approach we will still observe readiness oscillating/flapping, but container restarts should be less frequent.

Reply:
:rocket:

Reply:
the reasoning is correct, like the idea. Willing to see how it works in a real scenario 🍿

Reply:
Tested on ginger. Compared 1.6.10 and this PR. Fixed HPA to 2 replicas and reduced max-concurrent-connections to 30 to make it easy to saturate connections.

With nginx overloaded 1.6.10 replicas were restarting:
```
{⎈ |giantswarm-t0d0x:default} ➜  ~ k get pods -n kube-system -l k8s-app=nginx-ingress-controller -w
NAME                                      READY   STATUS    RESTARTS   AGE
nginx-ingress-controller-bc9844cd-sg4zh   1/1     Running   0          50s
nginx-ingress-controller-bc9844cd-znc58   1/1     Running   0          50s
nginx-ingress-controller-bc9844cd-sg4zh   0/1     Running   0          4m17s
nginx-ingress-controller-bc9844cd-sg4zh   0/1     Running   1          4m43s
nginx-ingress-controller-bc9844cd-sg4zh   1/1     Running   1          4m47s
nginx-ingress-controller-bc9844cd-sg4zh   0/1     Running   1          10m
nginx-ingress-controller-bc9844cd-sg4zh   0/1     Running   2          10m
nginx-ingress-controller-bc9844cd-sg4zh   1/1     Running   2          10m
nginx-ingress-controller-bc9844cd-znc58   0/1     Running   0          11m
nginx-ingress-controller-bc9844cd-znc58   1/1     Running   0          11m
nginx-ingress-controller-bc9844cd-znc58   0/1     Running   0          12m
nginx-ingress-controller-bc9844cd-sg4zh   0/1     Running   2          12m
nginx-ingress-controller-bc9844cd-znc58   0/1     Running   1          12m
nginx-ingress-controller-bc9844cd-znc58   1/1     Running   1          12m
nginx-ingress-controller-bc9844cd-sg4zh   0/1     Running   3          12m
nginx-ingress-controller-bc9844cd-sg4zh   1/1     Running   3          12m
nginx-ingress-controller-bc9844cd-znc58   0/1     Running   1          14m
nginx-ingress-controller-bc9844cd-znc58   0/1     Running   2          14m
nginx-ingress-controller-bc9844cd-znc58   1/1     Running   2          14m
```

```
Status code distribution:
  [200]	587 responses
  [500]	43579 responses
  [502]	41773 responses
```

This PR based nginx would as expected have readiness fail from time to time due to overload. Replicas would recover in very next readiness check (done every 10sec). Liveness didn't fail, so no restarts and as a result more successfully processed requests:

```
{⎈ |giantswarm-t0d0x:default} ➜  ~ k get pods -n kube-system -l k8s-app=nginx-ingress-controller -w
NAME                                        READY   STATUS    RESTARTS   AGE
nginx-ingress-controller-6d7bfc54d7-k78w6   1/1     Running   0          58s
nginx-ingress-controller-6d7bfc54d7-mxgff   1/1     Running   0          58s
nginx-ingress-controller-6d7bfc54d7-k78w6   0/1     Running   0          2m17s
nginx-ingress-controller-6d7bfc54d7-k78w6   1/1     Running   0          2m27s
nginx-ingress-controller-6d7bfc54d7-mxgff   0/1     Running   0          3m20s
nginx-ingress-controller-6d7bfc54d7-mxgff   1/1     Running   0          3m30s
nginx-ingress-controller-6d7bfc54d7-k78w6   0/1     Running   0          3m57s
nginx-ingress-controller-6d7bfc54d7-k78w6   1/1     Running   0          4m7s
nginx-ingress-controller-6d7bfc54d7-mxgff   0/1     Running   0          4m30s
nginx-ingress-controller-6d7bfc54d7-mxgff   1/1     Running   0          4m40s
nginx-ingress-controller-6d7bfc54d7-k78w6   0/1     Running   0          5m7s
nginx-ingress-controller-6d7bfc54d7-k78w6   1/1     Running   0          5m17s
nginx-ingress-controller-6d7bfc54d7-k78w6   0/1     Running   0          5m47s
nginx-ingress-controller-6d7bfc54d7-k78w6   1/1     Running   0          5m57s
```

```
Status code distribution:
  [200]	712 responses
  [500]	28961 responses
  [502]	54750 responses
```